### PR TITLE
String calculator

### PR DESCRIPTION
**Pull request by Tamer-Assaliya**

### CODE REVIEW

File: StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+using Xunit;
+
+namespace StringCalculatorKata.Tests
+{
+    public class StringCalculatorShould
+    {
+        [Theory]
+        [InlineData("", 0)]
```
Comment (RandaZraik) [inline StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:9]: add a test case for null as well

File: StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        [InlineData("1", 1)]
+        [InlineData("1,2", 3)]
+        [InlineData("1\n2,3", 6)]
```
Comment (RandaZraik) [inline StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: what about if the string has non-integers? what's the result of `add`? it will throw an exception ;)

File: StringCalculatorKata/StringCalculatorKata/StringCalculator.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        public int add(string numbers)
+        {
+            _numbers = numbers;
+            _checkForCustumDelimiter();
+            if (_numbers == "")
+                return 0;
+            return _numbers.Split(_delimiters.ToArray()).Select((num) => int.Parse(num)).Sum();
+        }
+
+        private void _checkForCustumDelimiter()
```
Comment (RandaZraik) [inline StringCalculatorKata/StringCalculatorKata/StringCalculator.cs]: we dont usually add _ at the beginning of func names: `CheckForCustumDelimiter`

File: StringCalculatorKata/StringCalculatorKata/StringCalculator.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+using System.Collections.Generic;
+using System.Collections;
+using System.Linq;
+
+namespace StringCalculatorKata
+{
+    public class StringCalculator
+    {
+        private string _numbers;
+        private List<char> _delimiters = new List<char>() { ',', '\n' };
```
Comment (RandaZraik) [inline StringCalculatorKata/StringCalculatorKata/StringCalculator.cs]: this list should be const, list do it as an array instead of list

File: StringCalculatorKata/StringCalculatorKata/StringCalculator.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+            return _numbers.Split(_delimiters.ToArray()).Select((num) => int.Parse(num)).Sum();
+        }
+
+        private void _checkForCustumDelimiter()
+        {
+            bool isDelimmeterSpecified = _numbers.StartsWith("//");
+            if (isDelimmeterSpecified == true)
+            {
+                char customDelimiter = _numbers.Substring(2, 1).ToCharArray()[0];
+                _delimiters.Add(customDelimiter);
```
Comment (RandaZraik) [inline StringCalculatorKata/StringCalculatorKata/StringCalculator.cs]: You shouldn't add these custom delimiters to the `_delimiters` list. This list should have the default (always allowed) delimiters. The custom delimiters you find here are only specific to the input you calculate.

File: StringCalculatorKata/StringCalculatorKata/StringCalculator.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+namespace StringCalculatorKata
+{
+    public class StringCalculator
+    {
+        private string _numbers;
+        private List<char> _delimiters = new List<char>() { ',', '\n' };
+
+        public int add(string numbers)
+        {
+            _numbers = numbers;
```
Comment (RandaZraik) [inline StringCalculatorKata/StringCalculatorKata/StringCalculator.cs]: `Add` shouldn't save the input anywhere really. This method should be more abstract, and you can make it as a static method actually, no need to initialize an object from `StringCalculator` to use the methods it has.

File: StringCalculatorKata/StringCalculatorKata/StringCalculator.cs (hunk 5 above)
Comment (RandaZraik) [reply StringCalculatorKata/StringCalculatorKata/StringCalculator.cs]: `StringCalculator` should be more abstract and not store data really.

File: StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs (hunk 2 above)
Comment (RandaZraik) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: and let's also think about short inputs, such as: "//", "//1", ...

File: StringCalculatorKata/StringCalculatorKata/StringCalculator.cs (hunk 6 above)
Comment (Tamer-Assaliya) [reply StringCalculatorKata/StringCalculatorKata/StringCalculator.cs]: I made some changes and pushed them
you can check now ;)

File: StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs (hunk 2 above)
Comment (Tamer-Assaliya) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: Okey I only followed your instruct "Make sure you only test for correct inputs. There is no need to test for invalid inputs for this kata."
but I'll handle them ;)
Should/could I use try catch for the exceptions in general?

Comment (RandaZraik) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: ah makes sense.

What kind of exceptions? Because some exceptions shouldn't be thrown and we need to fix the code to avoid them.

Comment (RandaZraik) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: Like if you try to parse a non-integer input.

Comment (Tamer-Assaliya) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: aha Tamam that's in the next req.

Comment (RandaZraik) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: nice work, thanks tamer!!
Can you add test cases for these cases: (i bit you they will throw some exceptions ;) )

"6,1,1,1000,,2000"
"10001,\n2000"
"10001,a,2000"
"10001,,2000"
","

Comment (Tamer-Assaliya) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: Thank you! 
I almost forgot about these general exceptions 😅
I'll add them of course ;)

Comment (Tamer-Assaliya) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: Done
you can review the pushed code

Comment (RandaZraik) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: Thanks Tamer! And general note, you need the code to handle the exceptions to NOT throw them, so if you pass a char not an integer, then you shouldn't parse it and just throw an error that 'the input is invalid'.
You should catch these exceptions from code and handle them, or the program will break.

Comment (Tamer-Assaliya) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: Done..
I handled every invalid format exception and make the code returns 0 in that case.
Does that work well?

Comment (RandaZraik) [reply StringCalculatorKata/StringCalculatorKata.Tests/StringCalculatorShould.cs:12]: Fair enough :D
